tony-history-server/app/utils/HdfsUtils.java

```diff
@@ -63,27 +59,52 @@ static String contentOfHdfsFile(FileSystem fs, Path filePath) {
   }
 
   /**
-   * List all metadata file paths in {@code histFolder}.
+   * Extract the job id portion from the file path string.
+   * @param path file path string.
+   * @return job id
+   */
+  public static String getJobId(String path) {
+    if (path.length() == 0) return "";
+    String[] folderLayers = path.split("/");
+    return folderLayers[folderLayers.length - 1];
+  }
+
+  /**
+   * Check to see if the path {@code p} is a job folder path from the given
+   * folder name pattern {@code regex}
+   * @param p Path object
+   * @param regex regular expression string
+   * @return true if path {@code p} is a job folder path, false otherwise.
+   */
+  @VisibleForTesting
+  static boolean isJobFolder(Path p, String regex) {
+    return getJobId(p.toString()).matches(regex);
+  }
+
+  /**
+   * Find all job folders under {@code curr} that matches {@code regex} pattern
+   * and return a list of corresponding Path objects.
    * @param fs FileSystem object.
-   * @param histFolder full path of the history folder.
-   * @return A list of metadata Path objects.
+   * @param curr folder location Path object.
+   * @param regex regular expression string.
+   * @return list of job Path objects in {@code curr} folder.
    */
-  public static List<Path> getMetadataFilePaths(FileSystem fs, String histFolder) {
-    List<Path> paths = new ArrayList<>();
-    FileStatus[] lsHist;
+  public static List<Path> getJobFolders(FileSystem fs, Path curr, String regex) {
+    List<Path> intermediateFolders = new ArrayList<>();
+    if (isJobFolder(curr, regex)) {
+      intermediateFolders.add(curr);
+      return intermediateFolders;
+    }
     try {
-      lsHist = fs.listStatus(new Path(histFolder));
+      intermediateFolders = Arrays.stream(fs.listStatus(curr))
+          .filter(FileStatus::isDirectory)
+          .map(fileStatus -> getJobFolders(fs, fileStatus.getPath(), regex))
+          .flatMap(List::stream)
+          .collect(Collectors.toList());
     } catch (IOException e) {
-      LOG.error("Failed to read history folder", e);
-      return paths;
-    }
-
-    for (FileStatus nodes : lsHist) {
-      if (nodes.isDirectory()) {
-        paths.add(new Path(nodes.getPath().toString() + "/metadata.json"));
-      }
+      LOG.error("Failed to traverse down history folder");
```

Comment:
I think I just want to return the `intermediateFolders`, assuming that it would just be an empty list when `fs.listStatus` throws error. If my assumption is incorrect, then I'll just return an empty list.

Reply:
So I did some googling. It seems like Java defaults back to what the variable is set to before the exception is thrown. In this case, `intermediateFolders` would be set to an empty `ArrayList`, and that was my assumption. 
I'll add the `e` the `LOG` call. Fixed: https://github.com/linkedin/TonY/pull/81/commits/ac6fbe127274f0b218a87dd44b97ca0c7c112d57.